doc/ppx-for-plugin-authors.rst

```diff
@@ -0,0 +1,101 @@
+**********************
+PPX for plugin authors
+**********************
+
+This section describes how to use ``ppxlib`` for PPX plugin authors.
+
+Metaquot
+------------
+
+``metaquot`` is a PPX plugin to help you write PPX plugins. It lets you write AST node values
+using the actual OCaml syntax instead of building them with the more verbose AST types or
+``Ast_builder``.
+
+To use metaquot you need to add it to the list of preprocessor for your PPX plugin:
+
+.. code:: scheme
+
+          (library
+           (name my_plugin_lib)
+           (preprocess (pps ppxlib.metaquot)))
+
+Metaquot exposes different extensions. The extension you should use depends on the type of AST node
+you're trying to write. You can use the following extensions with the following syntax:
+- ``expr`` for ``Parsetree.expression``: ``[%expr 1 + 1]``
+- ``pat`` for ``Parsetree.pattern``: ``[%pat? ("", _)]``
+- ``type`` for ``Parsetree.core_type``: ``[%type: int -> string]``
+- ``stri`` for ``Parsetree.structure_item``: ``[%str let a = 1]``
+- ``sigi`` for ``Parsetree.signature_item``: ``[%sigi: val i : int]``
+- ``str`` and ``sig`` respectively for ``Parsetree.structure`` and ``Parsetree.signature``. They use
+  similar syntax to the ``_item`` extensions above as they are just a list of such items.
+
+If you consider the first example ``[%expr 1 + 1]``, metaquot will actually expand it into:
+
+.. code:: ocaml
+
+          {
+            pexp_desc =
+              (Pexp_apply
+                 ({
+                    pexp_desc = (Pexp_ident { txt = (Lident "+"); loc });
+                    pexp_loc = loc;
+                    pexp_attributes = []
+                  },
+                   [(Nolabel,
+                      {
+                        pexp_desc = (Pexp_constant (Pconst_integer ("1", None)));
+                        pexp_loc = loc;
+                        pexp_attributes = []
+                      });
+                   (Nolabel,
+                     {
+                       pexp_desc = (Pexp_constant (Pconst_integer ("1", None)));
+                       pexp_loc = loc;
+                       pexp_attributes = []
+                     })]));
+            pexp_loc = loc;
+            pexp_attributes = []
+          }
+
+For this to compile you need the AST types to be in the scope so you should always use metaquot
+where ``Ppxlib`` is opened.
+You'll also note that the generated node expects a ``loc : Location.t`` value to be available. The
+produced AST node value and every other nodes within it will be located to ``loc``. You should make
+sure ``loc`` is the location you want for your generated code when using metaquot.
+
+Using only these extensions you can only produce constant/static AST nodes. Metaquot as a solution
+for that as well: quasi-quotation. You can use quasi-quotation to insert any expression representing
```

Comment:
It is the first time I see the term quasi-quotation. Is is standard in this context? I always called these antiquotations

Reply:
I also usually use anti-quotation but I think I read the term quasi-quotation in this [whitequark's article](https://whitequark.org/blog/2014/04/16/a-guide-to-extension-points-in-ocaml/) the other day and got confused.

By re-reading it now it seems clear that quasiquotation is in fact used to designate `metaquot` as a whole which makes much more sense.

I'll change that!